lib/active_publisher/configuration.rb

```diff
@@ -1,23 +1,86 @@
+require "yaml"
+
 module ActivePublisher
   class Configuration
-    attr_accessor :default_exchange,
-                  :error_handler,
+    attr_accessor :async_publisher,
+                  :async_publisher_drop_messages_when_queue_full,
+                  :async_publisher_max_queue_size,
+                  :async_publisher_supervisor_interval,
+                  :async_publisher_error_handler,
                   :heartbeat,
                   :host,
                   :hosts,
+                  :password,
                   :port,
                   :publisher_confirms,
-                  :timeout
+                  :seconds_to_wait_for_graceful_shutdown,
+                  :username,
+                  :timeout,
+                  :virtual_host
+
+    CONFIGURATION_MUTEX = ::Mutex.new
+
+    DEFAULTS = {
+      :async_publisher => 'memory',
+      :async_publisher_drop_messages_when_queue_full => false,
+      :async_publisher_max_queue_size => 1_000_000,
+      :async_publisher_supervisor_interval => 200, # in milliseconds
+      :heartbeat => 5,
+      :host => 'localhost',
+      :hosts => [],
+      :port => 5672,
+      :publisher_confirms => false,
+      :seconds_to_wait_for_graceful_shutdown => 30,
+      :timeout => 1,
+      :username => "guest",
+      :password => "guest",
+      :virtual_host => "/"
+    }
+
+    ##
+    # Class Methods
+    #
+    def self.configure_from_yaml_and_cli(cli_options = {}, reload = false)
```

Comment:
This method is pretty large. Perhaps we could take this opportunity to split it up?